akka-docs/src/main/paradox/cluster-usage.md

```diff
@@ -12,6 +12,53 @@ To use Akka Cluster, you must add the following dependency in your project:
+a single responsibility as identified by Jonas Bonér in the book
+[Reactive Microsystems: The Evolution of Microservices at Scale](https://info.lightbend.com/ebook-reactive-microservices-the-evolution-of-microservices-at-scale-register.html).
+
+In a microservices architecture, you should consider communication within a service and between services.
+
+In general we recommend against using Akka Cluster and actor messaging between _different_ services because that
+would result in a too tight code coupling between the services and difficulties deploying these independent of
+each other, which is one of the main reasons for using a microservices architecture.
+See the discussion on
+@scala[[Internal and External Communication](https://www.lagomframework.com/documentation/current/scala/InternalAndExternalCommunication.html)]
+@java[[Internal and External Communication](https://www.lagomframework.com/documentation/current/java/InternalAndExternalCommunication.html)]
+in the docs of the [Lagom Framework](https://www.lagomframework.com) (where each microservice is an Akka Cluster)
+for some background on this.
+
+Nodes of a single service (collectively called a cluster) require less decoupling. They share the same code and
+are managed together, as a set, by a single team or individual. For this reason, intra-service communication
+can take advantage of Akka Cluster, failure management and actor messaging, which is convenient to use and has great
+performance.
+
+Between different services [Akka HTTP](https://doc.akka.io/docs/akka-http/current) or
+[Akka gRPC](https://developer.lightbend.com/docs/akka-grpc/current/) can be used for synchronous (yet non-blocking)
+communication and [Akka Streams Kafka](https://doc.akka.io/docs/akka-stream-kafka/current/home.html) or other
+[Alpakka](https://developer.lightbend.com/docs/alpakka/current/) connectors for integration asynchronous communication.
+All those communication mechanisms work well with streaming of messages with end-to-end back-pressure, and the
+synchronous communication tools can also be used for single request response interactions. It is also important
+to note that when using these tools both sides of the communication do not have to be implemented with Akka,
+nor does the programming language matter.
+
+### Traditional distributed application
+
+We acknowledge that microservices also introduce many new challenges and it's not the only way to
+build applications. A traditional distributed application may have less complexity and work well in many cases.
+For example for a small startup, with a single team, building an application where time to market is everything.
+Akka Cluster can efficiently be used for building such distributed application. Even though it is considered an
+anti-pattern in the microservices world, it's a great pattern if you have the right use-cases. It has served
+the industry and many Akka users well for years and is still a valid choice.
```

Comment:
I found this section confusing. What is the anti-pattern? Building a clustered microservices or building a monolith app when a startup needs to quickly bring something to market?

Reply:
the section is about traditional distributed application so __it__ refers to that, but I can clarify by repeating one more time: "Even though it is considered anti-pattern in the microservices world to build a tightly coupled distributed application, "

Reply:
I think there are three different patterns: two "pro" patterns and one anti-pattern:

### Microservices

These are independently-deployable units that are built from separate code trees—possibly from a shared repository, possibly separate, but either way with some kind of code boundary. Because their code is independent, they should only communicate with each other using loosely-coupled protocols and should not share a cluster. This is to retain the autonomy of the different units, which is a principal goal of microservices.

Within each service unit, individual instances of that unit can form a cluster with each other to have a tighter form of communication. This works well because the code is shared and deployed together. There might be two versions running concurrently during a rolling deployment, but deployment of the entire set has a single point of control.

### Distributed monolith

This is the anti-pattern. You have multiple services that are built and deployed independently from each other, but they have a tight coupling that makes this very risky, such as a shared cluster, or a shared database schema. There is a false sense of autonomy because of the physical separation of the code and deployment units, but you are likely to encounter problems because of changes in the implementation of one service leaking into the behavior of others. Organizations that find themselves in this situation often react by trying to centrally coordinate deployment of multiple services, at which point you have lost the principal benefit of microservices while taking on the costs. You are in a halfway state with things that aren't really separable being built and deployed in a separate way. Some people do this, and some manage to make it work, but it's not something we would recommend (IMHO) and it needs to be carefully managed.

### Traditional distributed applications

In this case, you have a single deployment unit, built from a single code base (or using traditional binary dependency management to modularize) but deployed across many nodes using a single cluster (maybe a lot more than individual microservice would be). Tighter coupling is OK, because there is a central point of deployment and control. In some cases, nodes may have specialized runtime roles which means that the cluster is not totally homogenous (e.g., "front-end" and "back-end" nodes, or dedicated master/worker nodes) but if these are run from the same built artifacts this is just a runtime behavior and doesn't cause the same kind of problems you might get from tight coupling of totally separate artifacts. This has more in common with the microservices architecture than the distributed monolith—the recommendations are the same, it's only the granularity of the deployment units that are different.


I think the second and third are very different cases, and the goal of this should be to help people understand the difference and that we discourage the second while accepting the third.

Reply:
@TimMoore Good point to make that clear. I have tried to work in some of your text, please take another look.